financial_calculations/coupon_rates.py

```diff
@@ -0,0 +1,77 @@
+import numpy as np
+import pandas as pd
+from datetime import datetime, timedelta
+from dateutil.relativedelta import relativedelta
+from scipy.optimize import minimize
+from utils import (
+    discount_cash_flows
+)
+from financial_calculations.bond_cash_flows import (
+    create_semi_bond_cash_flows
+)
+
+def calculate_coupon_rate(start_date, maturity_years, par_value, forward_curve, initial_guess=0.04):
```

Comment:
no need for a search here

$(C \sum_{t=1}^T .5 * disc_t) + 100* disc_T = 100 * disc_0$

solve for $C$.  Time $t$ represents payment dates